net_util/src/queue_pair.rs

```diff
@@ -271,8 +273,35 @@ impl NetQueuePair {
             .ok_or(NetQueuePairError::NoMemoryConfigured)
             .map(|m| m.memory())?;
 
-        self.tx
-            .process_desc_chain(&mem, &mut self.tap, &mut queue, &mut self.tx_rate_limiter)?;
+        let tx_tap_retry = self.tx.process_desc_chain(
+            &mem,
+            &mut self.tap,
+            &mut queue,
+            &mut self.tx_rate_limiter,
+        )?;
+
+        // We got told to try again when writing to the tap. Wait for the TAP to be writable
+        if tx_tap_retry && !self.tx_tap_listening {
+            register_listener(
+                self.epoll_fd.unwrap(),
+                self.tap.as_raw_fd(),
+                epoll::Events::EPOLLOUT,
+                u64::from(self.tap_tx_event_id),
+            )
+            .map_err(NetQueuePairError::UnregisterListener)?;
+            self.tx_tap_listening = true;
+            info!("Writing to TAP returned EAGAIN. Listening for TAP to become writable.");
+        } else if self.tx_tap_listening {
```

Comment:
I might be wrong here, but I think this should also check to make sure there was no EAGAIN. Right now, if a listener is set up, it will be removed on the next call of this method. In practice, from the logs I saw (prior to running this patch), this function might run multiple times in quick succession where each invocation hits EAGAIN. In that scenario, the second call would remove the listener again. Likewise, afaik there is no guarantee that a tap becoming writeable again is able to send all the pending data, yet the listener would be removed as well.

Reply:
@LeeSmet I actually tried to handle the successive EAGAIN hence the `else if`. If `process_desc_chain` encounters EAGAIN then it returns true (`tx_tap_retry`).

If we get told we need to retry and no listener is set then add the listener (`if tx_tap_retry && !self.tx_tap_listening`) if we don't need to retry and a listener is set ` } else if self.tx_tap_listening {` remove the listener.

So if we have to retry and we already have the listener then none of the above code will be run and the listener retained.

Please be aware that the `else if` is not re-evalulated. So this is equivalent to.

`if tx_tap_retry && !self.tx_tap_listening { .. } else { if self.tx_tap_listening { .. }  }` however clippy will complain about this style.

Reply:
Oops. Scratch that. I see I was missing an extra clause on the `} else if {`

Reply:
Thanks! :-)